# TRAIN CNN
## Baby- Not Baby

In [1]:
# LOCAL PATHS
img_path = '../scratch/images/ne_aclab/'


In [2]:
train_path = img_path + 'train/'
test_path = img_path + 'validate/'

In [3]:
### SET MODEL TO RUN HERE!! ###

run_model = 3

### SET MODEL TO RUN HERE!! ###

In [4]:
import os

import keras
from keras.layers import Dense, Flatten
from keras.models import Sequential
from keras.utils import to_categorical

from keras.preprocessing.image import ImageDataGenerator
from keras.layers import Conv2D, MaxPool2D, Dropout
from keras.applications import MobileNetV2, VGG16, InceptionV3, Xception, ResNet50
from keras.callbacks import EarlyStopping

from keras.layers import GlobalAveragePooling2D


In [5]:
datagen = ImageDataGenerator(rescale=1. / 255)

In [6]:
train_aug = datagen.flow_from_directory(
    train_path,
    target_size=(299, 299),
    class_mode=None,  
    shuffle=False
)


Found 3000 images belonging to 2 classes.


In [7]:
test_aug = datagen.flow_from_directory(
    test_path,
    target_size=(299, 299),
    class_mode=None,  
    shuffle=False
)


Found 822 images belonging to 2 classes.


In [8]:
print(train_aug.class_indices)
print(test_aug.class_indices)


{'baby': 0, 'notbaby': 1}
{'baby': 0, 'notbaby': 1}


In [9]:
if(run_model == 1):
    conv_base = Xception(include_top=False, input_shape=(299, 299, 3))
elif(run_model == 2):
    conv_base = MobileNetV2(include_top=False, input_shape=(299, 299, 3))
elif(run_model == 3):
    conv_base = InceptionV3(include_top=False, input_shape=(299, 299, 3))
elif(run_model == 4):
    conv_base = ResNet50(include_top=False, input_shape=(299, 299, 3))
elif(run_model == 5):
    conv_base = VGG16(include_top=False, input_shape=(299, 299, 3))

print(f'Running Model: {run_model}')

Running Model: 3


In [10]:
X_train = conv_base.predict(train_aug)
X_test = conv_base.predict(test_aug)


In [11]:
y_train = train_aug.labels 
y_train_binarized = to_categorical(y_train)

y_test = test_aug.labels
y_test_binarized = to_categorical(y_test)

In [12]:
model = Sequential()

model.add(GlobalAveragePooling2D())

model.add(Dense(50, activation = 'relu'))
model.add(Dense(350, activation = 'relu'))
model.add(Dense(150, activation = 'relu'))
model.add(Dense(50, activation = 'relu'))

model.add(Dense(2, activation='softmax'))
model.compile(
    loss='categorical_crossentropy', 
    optimizer='adam', 
    metrics=['accuracy', 'Precision', 'Recall']
)


In [13]:
callback = EarlyStopping(patience=5)

history = model.fit(
    X_train, 
    y_train_binarized, 
    epochs = 20,
    validation_data=(X_test, y_test_binarized),
    callbacks=[callback],
    verbose=1
)

print(f'Model Ran: {run_model}')

Epoch 1/20
94/94 [==============================] - 3s 19ms/step - loss: 0.2725 - accuracy: 0.8591 - precision: 0.8591 - recall: 0.8591 - val_loss: 0.0643 - val_accuracy: 0.9793 - val_precision: 0.9793 - val_recall: 0.9793
Epoch 2/20
94/94 [==============================] - 1s 6ms/step - loss: 0.0376 - accuracy: 0.9860 - precision: 0.9860 - recall: 0.9860 - val_loss: 0.1387 - val_accuracy: 0.9586 - val_precision: 0.9586 - val_recall: 0.9586
Epoch 3/20
94/94 [==============================] - 1s 7ms/step - loss: 0.0668 - accuracy: 0.9747 - precision: 0.9747 - recall: 0.9747 - val_loss: 0.1127 - val_accuracy: 0.9720 - val_precision: 0.9720 - val_recall: 0.9720
Epoch 4/20
94/94 [==============================] - 1s 7ms/step - loss: 0.0093 - accuracy: 0.9972 - precision: 0.9972 - recall: 0.9972 - val_loss: 0.2170 - val_accuracy: 0.9477 - val_precision: 0.9477 - val_recall: 0.9477
Epoch 5/20
94/94 [==============================] - 1s 7ms/step - loss: 0.0461 - accuracy: 0.9849 - precision: 

In [14]:
print(f'Model Ran: {run_model}')

Model Ran: 3


In [15]:
if not os.path.exists('../models'):
    os.mkdir('../models')
save_path = f'../models/toddler_tracker_model_{run_model}.h5'

In [16]:
# model.save(save_path)
# print(f'File saved to: {save_path}')

In [17]:
# from IPython.lib.display import Audio
# import numpy as np

# framerate = 4410
# play_time_seconds = 2

# t = np.linspace(0, play_time_seconds, framerate*play_time_seconds)
# audio_data = np.sin(2*np.pi*110*t)# + np.sin(2*np.pi*78*t)
# Audio(audio_data, rate=framerate, autoplay=True)

# RESULTS

<!--
## 0. None
### val_accuracy: 0.5000 - val_precision: 0.5000 - val_recall: 0.5000
<details><summary>Epoch 1/20</summary>


</details>

`Epoch`

`9494`

| Type     | With | Without |
| -------- |:----:|:-------:|
| Callback |  5   |         |
| Dropout  | 0.2  |         |
| l2       | 0.1  |         |
-_>

## 1. Xception
### val_accuracy: 0.9611 - val_precision: 0.9611 - val_recall: 0.9611
<details><summary>Epoch 1/20</summary>

`94/94 [==============================] - 3s 23ms/step - loss: 0.2476 - accuracy: 0.8995 - precision: 0.8995 - recall: 0.8995 - val_loss: 0.0338 - val_accuracy: 0.9878 - val_precision: 0.9878 - val_recall: 0.9878`

`Epoch 2/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.0217 - accuracy: 0.9914 - precision: 0.9914 - recall: 0.9914 - val_loss: 0.0293 - val_accuracy: 0.9903 - val_precision: 0.9903 - val_recall: 0.9903`

`Epoch 3/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.0150 - accuracy: 0.9948 - precision: 0.9948 - recall: 0.9948 - val_loss: 0.0382 - val_accuracy: 0.9854 - val_precision: 0.9854 - val_recall: 0.9854`

`Epoch 4/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.0070 - accuracy: 0.9964 - precision: 0.9964 - recall: 0.9964 - val_loss: 0.2745 - val_accuracy: 0.9489 - val_precision: 0.9489 - val_recall: 0.9489`

`Epoch 5/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.0160 - accuracy: 0.9945 - precision: 0.9945 - recall: 0.9945 - val_loss: 0.0762 - val_accuracy: 0.9830 - val_precision: 0.9830 - val_recall: 0.9830`

`Epoch 6/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.0033 - accuracy: 0.9990 - precision: 0.9990 - recall: 0.9990 - val_loss: 0.1230 - val_accuracy: 0.9793 - val_precision: 0.9793 - val_recall: 0.9793`


</details>

`Epoch 7/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.0011 - accuracy: 0.9998 - precision: 0.9998 - recall: 0.9998 - val_loss: 0.2759 - val_accuracy: 0.9611 - val_precision: 0.9611 - val_recall: 0.9611`



## 2. MobileNetV2
### val_accuracy: 0.9878 - val_precision: 0.9878 - val_recall: 0.9878
<details><summary>Epoch 1/20</summary>

`94/94 [==============================] - 3s 20ms/step - loss: 0.2429 - accuracy: 0.9008 - precision: 0.9008 - recall: 0.9008 - val_loss: 0.0418 - val_accuracy: 0.9878 - val_precision: 0.9878 - val_recall: 0.9878`

`Epoch 2/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0267 - accuracy: 0.9913 - precision: 0.9913 - recall: 0.9913 - val_loss: 0.1128 - val_accuracy: 0.9659 - val_precision: 0.9659 - val_recall: 0.9659`

`Epoch 3/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0138 - accuracy: 0.9956 - precision: 0.9956 - recall: 0.9956 - val_loss: 0.0426 - val_accuracy: 0.9878 - val_precision: 0.9878 - val_recall: 0.9878`

`Epoch 4/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0060 - accuracy: 0.9987 - precision: 0.9987 - recall: 0.9987 - val_loss: 0.0442 - val_accuracy: 0.9830 - val_precision: 0.9830 - val_recall: 0.9830`

`Epoch 5/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0171 - accuracy: 0.9951 - precision: 0.9951 - recall: 0.9951 - val_loss: 0.1015 - val_accuracy: 0.9769 - val_precision: 0.9769 - val_recall: 0.9769`

</details>

`Epoch 6/20`

`94/94 [==============================] - 1s 7ms/step - loss: 9.8219e-04 - accuracy: 0.9999 - precision: 0.9999 - recall: 0.9999 - val_loss: 0.0699 - val_accuracy: 0.9878 - val_precision: 0.9878 - val_recall: 0.9878`


## 3. InceptionV3
### val_accuracy: 0.9854 - val_precision: 0.9854 - val_recall: 0.9854
<details><summary>Epoch 1/20</summary>

`94/94 [==============================] - 3s 21ms/step - loss: 0.3015 - accuracy: 0.8544 - precision: 0.8544 - recall: 0.8544 - val_loss: 0.0726 - val_accuracy: 0.9720 - val_precision: 0.9720 - val_recall: 0.9720`

`Epoch 2/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0393 - accuracy: 0.9875 - precision: 0.9875 - recall: 0.9875 - val_loss: 0.1417 - val_accuracy: 0.9562 - val_precision: 0.9562 - val_recall: 0.9562`

`Epoch 3/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0364 - accuracy: 0.9871 - precision: 0.9871 - recall: 0.9871 - val_loss: 0.0466 - val_accuracy: 0.9866 - val_precision: 0.9866 - val_recall: 0.9866`

`Epoch 4/20`

`94/94 [==============================] - 1s 8ms/step - loss: 0.0222 - accuracy: 0.9892 - precision: 0.9892 - recall: 0.9892 - val_loss: 0.1109 - val_accuracy: 0.9732 - val_precision: 0.9732 - val_recall: 0.9732`

`Epoch 5/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0184 - accuracy: 0.9928 - precision: 0.9928 - recall: 0.9928 - val_loss: 0.0605 - val_accuracy: 0.9781 - val_precision: 0.9781 - val_recall: 0.9781`

`Epoch 6/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0095 - accuracy: 0.9969 - precision: 0.9969 - recall: 0.9969 - val_loss: 0.1921 - val_accuracy: 0.9513 - val_precision: 0.9513 - val_recall: 0.9513`

`Epoch 7/20`

`94/94 [==============================] - 1s 8ms/step - loss: 0.0246 - accuracy: 0.9905 - precision: 0.9905 - recall: 0.9905 - val_loss: 0.1109 - val_accuracy: 0.9781 - val_precision: 0.9781 - val_recall: 0.9781`

</details>

`Epoch 8/20`

`94/94 [==============================] - 1s 7ms/step - loss: 0.0072 - accuracy: 0.9965 - precision: 0.9965 - recall: 0.9965 - val_loss: 0.0989 - val_accuracy: 0.9854 - val_precision: 0.9854 - val_recall: 0.9854`



## 4. ResNet50
### val_accuracy: 0.7676 - val_precision: 0.7676 - val_recall: 0.7676
<details><summary>Epoch 1/20</summary>

`94/94 [==============================] - 4s 27ms/step - loss: 0.6804 - accuracy: 0.5532 - precision: 0.5532 - recall: 0.5532 - val_loss: 0.6193 - val_accuracy: 0.6679 - val_precision: 0.6679 - val_recall: 0.6679`

`Epoch 2/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.6182 - accuracy: 0.6418 - precision: 0.6418 - recall: 0.6418 - val_loss: 0.5544 - val_accuracy: 0.7056 - val_precision: 0.7056 - val_recall: 0.7056`

`Epoch 3/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.5594 - accuracy: 0.7160 - precision: 0.7160 - recall: 0.7160 - val_loss: 0.5245 - val_accuracy: 0.7482 - val_precision: 0.7482 - val_recall: 0.7482`

`Epoch 4/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.5086 - accuracy: 0.7562 - precision: 0.7562 - recall: 0.7562 - val_loss: 0.8030 - val_accuracy: 0.5803 - val_precision: 0.5803 - val_recall: 0.5803`

`Epoch 5/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.4777 - accuracy: 0.7917 - precision: 0.7917 - recall: 0.7917 - val_loss: 0.5800 - val_accuracy: 0.7336 - val_precision: 0.7336 - val_recall: 0.7336`

`Epoch 6/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.4898 - accuracy: 0.7755 - precision: 0.7755 - recall: 0.7755 - val_loss: 0.4972 - val_accuracy: 0.7506 - val_precision: 0.7506 - val_recall: 0.7506`

`Epoch 7/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.4981 - accuracy: 0.7568 - precision: 0.7568 - recall: 0.7568 - val_loss: 0.6202 - val_accuracy: 0.6496 - val_precision: 0.6496 - val_recall: 0.6496`

`Epoch 8/20`

`94/94 [==============================] - 1s 12ms/step - loss: 0.4701 - accuracy: 0.7901 - precision: 0.7901 - recall: 0.7901 - val_loss: 0.5036 - val_accuracy: 0.7628 - val_precision: 0.7628 - val_recall: 0.7628`

`Epoch 9/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.4610 - accuracy: 0.7827 - precision: 0.7827 - recall: 0.7827 - val_loss: 0.5983 - val_accuracy: 0.7019 - val_precision: 0.7019 - val_recall: 0.7019`

`Epoch 10/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4502 - accuracy: 0.7924 - precision: 0.7924 - recall: 0.7924 - val_loss: 0.5182 - val_accuracy: 0.7470 - val_precision: 0.7470 - val_recall: 0.7470`

`Epoch 11/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4580 - accuracy: 0.7870 - precision: 0.7870 - recall: 0.7870 - val_loss: 0.4923 - val_accuracy: 0.7628 - val_precision: 0.7628 - val_recall: 0.7628`

`Epoch 12/20`

`94/94 [==============================] - 1s 11ms/step - loss: 0.4664 - accuracy: 0.7868 - precision: 0.7868 - recall: 0.7868 - val_loss: 0.7019 - val_accuracy: 0.5973 - val_precision: 0.5973 - val_recall: 0.5973`

`Epoch 13/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4421 - accuracy: 0.7956 - precision: 0.7956 - recall: 0.7956 - val_loss: 0.5109 - val_accuracy: 0.7652 - val_precision: 0.7652 - val_recall: 0.7652`

`Epoch 14/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4153 - accuracy: 0.8196 - precision: 0.8196 - recall: 0.8196 - val_loss: 0.5497 - val_accuracy: 0.7433 - val_precision: 0.7433 - val_recall: 0.7433`

`Epoch 15/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4656 - accuracy: 0.7892 - precision: 0.7892 - recall: 0.7892 - val_loss: 0.4800 - val_accuracy: 0.7676 - val_precision: 0.7676 - val_recall: 0.7676`

`Epoch 16/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4006 - accuracy: 0.8237 - precision: 0.8237 - recall: 0.8237 - val_loss: 0.6150 - val_accuracy: 0.6813 - val_precision: 0.6813 - val_recall: 0.6813`

`Epoch 17/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4215 - accuracy: 0.8001 - precision: 0.8001 - recall: 0.8001 - val_loss: 0.5112 - val_accuracy: 0.7713 - val_precision: 0.7713 - val_recall: 0.7713`

`Epoch 18/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4182 - accuracy: 0.8119 - precision: 0.8119 - recall: 0.8119 - val_loss: 0.6377 - val_accuracy: 0.6776 - val_precision: 0.6776 - val_recall: 0.6776`

`Epoch 19/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4042 - accuracy: 0.8143 - precision: 0.8143 - recall: 0.8143 - val_loss: 0.5822 - val_accuracy: 0.7068 - val_precision: 0.7068 - val_recall: 0.7068`


</details>

`Epoch 20/20`

`94/94 [==============================] - 1s 10ms/step - loss: 0.4088 - accuracy: 0.8087 - precision: 0.8087 - recall: 0.8087 - val_loss: 0.4906 - val_accuracy: 0.7676 - val_precision: 0.7676 - val_recall: 0.7676`


## 5. VGG16
### val_accuracy: 0.9465 - val_precision: 0.9465 - val_recall: 0.9465
<details><summary>Epoch 1/20</summary>

`94/94 [==============================] - 2s 17ms/step - loss: 0.5195 - accuracy: 0.7519 - precision: 0.7519 - recall: 0.7519 - val_loss: 0.3377 - val_accuracy: 0.8613 - val_precision: 0.8613 - val_recall: 0.8613`

`Epoch 2/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.2555 - accuracy: 0.8957 - precision: 0.8957 - recall: 0.8957 - val_loss: 0.3324 - val_accuracy: 0.8564 - val_precision: 0.8564 - val_recall: 0.8564`

`Epoch 3/20`

`94/94 [==============================] - 0s 4ms/step - loss: 0.2026 - accuracy: 0.9172 - precision: 0.9172 - recall: 0.9172 - val_loss: 0.3236 - val_accuracy: 0.8650 - val_precision: 0.8650 - val_recall: 0.8650`

`Epoch 4/20`

`94/94 [==============================] - 0s 4ms/step - loss: 0.2122 - accuracy: 0.9182 - precision: 0.9182 - recall: 0.9182 - val_loss: 0.2106 - val_accuracy: 0.9173 - val_precision: 0.9173 - val_recall: 0.9173`

`Epoch 5/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.1416 - accuracy: 0.9471 - precision: 0.9471 - recall: 0.9471 - val_loss: 0.2632 - val_accuracy: 0.9002 - val_precision: 0.9002 - val_recall: 0.9002`

`Epoch 6/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.1491 - accuracy: 0.9404 - precision: 0.9404 - recall: 0.9404 - val_loss: 0.2497 - val_accuracy: 0.9100 - val_precision: 0.9100 - val_recall: 0.9100`

`Epoch 7/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.1552 - accuracy: 0.9361 - precision: 0.9361 - recall: 0.9361 - val_loss: 0.2326 - val_accuracy: 0.9088 - val_precision: 0.9088 - val_recall: 0.9088`

`Epoch 8/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.0985 - accuracy: 0.9612 - precision: 0.9612 - recall: 0.9612 - val_loss: 0.2097 - val_accuracy: 0.9185 - val_precision: 0.9185 - val_recall: 0.9185`

`Epoch 9/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.0738 - accuracy: 0.9724 - precision: 0.9724 - recall: 0.9724 - val_loss: 0.1711 - val_accuracy: 0.9294 - val_precision: 0.9294 - val_recall: 0.9294`

`Epoch 10/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.0819 - accuracy: 0.9707 - precision: 0.9707 - recall: 0.9707 - val_loss: 0.1413 - val_accuracy: 0.9501 - val_precision: 0.9501 - val_recall: 0.9501`

`Epoch 11/20`

`94/94 [==============================] - 0s 4ms/step - loss: 0.0669 - accuracy: 0.9722 - precision: 0.9722 - recall: 0.9722 - val_loss: 0.2050 - val_accuracy: 0.9197 - val_precision: 0.9197 - val_recall: 0.9197`

`Epoch 12/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.0513 - accuracy: 0.9840 - precision: 0.9840 - recall: 0.9840 - val_loss: 0.1565 - val_accuracy: 0.9477 - val_precision: 0.9477 - val_recall: 0.9477`

`Epoch 13/20`

`94/94 [==============================] - 0s 4ms/step - loss: 0.0438 - accuracy: 0.9826 - precision: 0.9826 - recall: 0.9826 - val_loss: 0.1922 - val_accuracy: 0.9380 - val_precision: 0.9380 - val_recall: 0.9380`

`Epoch 14/20`

`94/94 [==============================] - 0s 3ms/step - loss: 0.0660 - accuracy: 0.9784 - precision: 0.9784 - recall: 0.9784 - val_loss: 0.1975 - val_accuracy: 0.9343 - val_precision: 0.9343 - val_recall: 0.9343`

</details>

`Epoch 15/20`

`94/94 [==============================] - 0s 4ms/step - loss: 0.0570 - accuracy: 0.9806 - precision: 0.9806 - recall: 0.9806 - val_loss: 0.1454 - val_accuracy: 0.9465 - val_precision: 0.9465 - val_recall: 0.9465`
